src/error.rs

```diff
@@ -18,7 +18,73 @@ impl fmt::Display for DatabaseConstraint {
 }
 
 #[derive(Debug, Error)]
-pub enum Error {
+/// The error types for database I/O, connection and query parameter
+/// construction.
+pub struct Error {
+    kind: ErrorKind,
+    original_code: Option<String>,
+    original_message: Option<String>,
+}
+
+pub(crate) struct ErrorBuilder {
+    kind: ErrorKind,
+    original_code: Option<String>,
+    original_message: Option<String>,
+}
+
+impl ErrorBuilder {
+    pub(crate) fn set_original_code(&mut self, code: impl Into<String>) -> &mut Self {
+        self.original_code = Some(code.into());
+        self
+    }
+
+    pub(crate) fn set_original_message(&mut self, message: impl Into<String>) -> &mut Self {
+        self.original_message = Some(message.into());
+        self
+    }
+
+    pub(crate) fn build(self) -> Error {
+        Error {
+            kind: self.kind,
+            original_code: self.original_code,
+            original_message: self.original_message,
+        }
+    }
+}
+
+impl Error {
+    pub(crate) fn builder(kind: ErrorKind) -> ErrorBuilder {
+        ErrorBuilder {
+            kind,
+            original_code: None,
+            original_message: None,
+        }
+    }
+
+    /// The error code sent by the database, if available.
+    pub fn original_code(&self) -> Option<&str> {
+        self.original_code.as_ref().map(|s| s.as_str())
+    }
+
+    /// The original error message sent by the database, if available.
+    pub fn original_message(&self) -> Option<&str> {
+        self.original_message.as_ref().map(|s| s.as_str())
+    }
+
+    /// A more specific error type for matching.
+    pub fn kind(&self) -> &ErrorKind {
+        &self.kind
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.kind.fmt(f)
+    }
+}
```

Comment:
no need to do it now, but we could include the original message and code (if present) in the error message too. But maybe that's confusing for people if they see that and expect prisma error codes.

Reply:
I don't really want that...